Should DInstanceD have a Maybe [DTyVarBndr] field?

This is the current definition of `DInstanceD`:

https://github.com/goldfirere/th-desugar/blob/e420eaeb80f108c186c26ea706c7d80ce5e3d106/Language/Haskell/TH/Desugar/AST.hs#L94

This makes sense, as it mirrors `InstanceD` in `template-haskell`. However, I'm wondering if the API might be made more complete by adding an additional field to `DInstanceD` of type `Maybe [DTyVarBndr]`. This idea was born out of a realization I had when I thought about instances like the one below (inspired by [GHC #16518](https://gitlab.haskell.org/ghc/ghc/issues/16518)):

```hs
class HTestEquality (f :: forall k. k -> Type) where
  hTestEquality :: forall k1 k2 (a :: k1) (b :: k2).
                   f a -> f b -> Maybe (a :~~: b)
newtype T :: (forall k. k -> Type) -> (forall k. k -> Type) where
  MkT :: forall (f :: forall k. k -> Type) k (a :: k). f a -> T f a

instance forall (f :: forall k. k -> Type). HTestEquality f => HTestEquality (T f) where
  hTestEquality (MkT x) (MkT y) = hTestEquality x y
```

How exactly would one represent `instance forall (f :: forall k. k -> Type). HTestEquality f => HTestEquality (T f)` in Template Haskell? There's not an obvious place to put the `forall (f :: forall k. k -> Type)` part, and if we omit it, then the instance will fail to typecheck. Luckily, there _is_ a way to accomplish this: use the following TH AST:

```
λ> ; $(newName "f" >>= \f -> newName "k" >>= \k -> pure [InstanceD Nothing [] (ForallT [KindedTV f (ForallT [PlainTV k] [] (ArrowT `AppT` VarT k `AppT` StarT))] [ConT ''HTestEquality `AppT` VarT f] (ConT ''HTestEquality `AppT` (ConT ''T `AppT` VarT f))) []])
<interactive>:33:5-254: Splicing declarations
    newName "f"
      >>=
        \ f_ahqw
          -> newName "k"
               >>=
                 \ k_ahqx
                   -> pure
                        [InstanceD
                           Nothing
                           []
                           (ForallT
                              [KindedTV
                                 f_ahqw
                                 (ForallT
                                    [PlainTV k_ahqx] [] (ArrowT `AppT` VarT k_ahqx `AppT` StarT))]
                              [ConT ''HTestEquality `AppT` VarT f_ahqw]
                              (ConT ''HTestEquality `AppT` (ConT ''T `AppT` VarT f_ahqw)))
                           []]
  ======>
    instance forall (f_ahrt :: forall k_ahru.
                               k_ahru -> Data.Kind.Type).
             HTestEquality f_ahrt => HTestEquality (T f_ahrt)
```

This trick is a bit grimy, but it has the nice property that we don't have to change the TH AST to accommodate the (relatively obscure) use case of higher-ranked kind variables. Instead, we only need make this change in the downstream `th-desugar` library by adding a field of type `Maybe [DTyVarBndr]` to `DInstanceD`. If that field is instantiated with `Just tvs`, then we use the trick above and explicitly quantify `tvs`. If it's `Nothing`, then proceed like we normally do.

If we extend `DInstanceD`, then we should also give `DStandaloneDerivD` the same treatment.

Bad news: this `forall` trick isn't supported on exactly one version of GHC that `th-desugar` currently supports: GHC 8.0.2.

```
Test/DsDec.hs:77:3: error:
    Malformed instance: forall a. S.ExCls (S.ExData a)
```

Less critically, using this trick with GHC 7.10.3 in a standalone `deriving` declaration causes GHC to freak out a bit:

```
[2 of 4] Compiling Dec              ( Test/Dec.hs, /home/rgscott/Documents/Hacking/Haskell/staging/th-desugar/dist-newstyle/build/x86_64-linux/ghc-7.10.3/th-desugar-1.10/t/spec/build/spec/spec-tmp/Dec.o )
Var/Type length mismatch: 
  [a_acV1]
  []
Var/Type length mismatch: 
  [a_acV1]
  []
[3 of 4] Compiling DsDec            ( Test/DsDec.hs, /home/rgscott/Documents/Hacking/Haskell/staging/th-desugar/dist-newstyle/build/x86_64-linux/ghc-7.10.3/th-desugar-1.10/t/spec/build/spec/spec-tmp/DsDec.o )
Var/Type length mismatch: 
  [a_acV1]
  []
Var/Type length mismatch: 
  [a_acV1]
  []
```

This is rather annoying, since it's not the oldest versions of GHC giving us problems, but rather versions that are smack-dab in the middle of the range of versions that `th-desugar` supports. Some possible remedies:

* We could prohibit using this trick unless GHC 8.2 or later is used, although that cuts off certain versions of GHC where this trick works perfectly well.
* Alternatively, we could throw an error when using this trick on the particular versions of GHC where it's known to break.
* Something else?

Thoughts?

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -26,6 +26,12 @@ Version 1.10
     now that each `DTySynEqn` contains the same `Name`.
   * There is now a field of type `Maybe [DTyVarBndr]` in the `DRuleP`
     constructor to represent bound type variables in `RULES` (if present).
+* Add a field of type `Maybe [DTyVarBndr]` to `DInstanceD` and
+  `DStandaloneDerivD` for optionally quantifying type variables explicitly.
+  If supplied with a `Just`, this sweetens the instance type to use a `ForallT`
+  to represent the explicit quantification. This trick is not supported for
+  `InstanceD` on GHC 8.0 and for `StandaloneDerivD` on GHC 7.10 or 8.0, so be
+  aware of this limitation if you supply `Just` for this field.
 * Add support for desugaring implicit params. This does not involve any changes
   to the `th-desugar` AST, as:
   * `(?x :: a) => ...` is desugared to `IP "x" a => ...`.

--- a/Language/Haskell/TH/Desugar/AST.hs
+++ b/Language/Haskell/TH/Desugar/AST.hs
@@ -91,7 +91,7 @@ data DDec = DLetDec DLetDec
           | DDataD NewOrData DCxt Name [DTyVarBndr] (Maybe DKind) [DCon] [DDerivClause]
           | DTySynD Name [DTyVarBndr] DType
           | DClassD DCxt Name [DTyVarBndr] [FunDep] [DDec]
-          | DInstanceD (Maybe Overlap) DCxt DType [DDec]
+          | DInstanceD (Maybe Overlap) (Maybe [DTyVarBndr]) DCxt DType [DDec]
           | DForeignD DForeign
           | DOpenTypeFamilyD DTypeFamilyHead
           | DClosedTypeFamilyD DTypeFamilyHead [DTySynEqn]
@@ -100,7 +100,7 @@ data DDec = DLetDec DLetDec
                        [DCon] [DDerivClause]
           | DTySynInstD DTySynEqn
           | DRoleAnnotD Name [Role]
-          | DStandaloneDerivD (Maybe DDerivStrategy) DCxt DType
+          | DStandaloneDerivD (Maybe DDerivStrategy) (Maybe [DTyVarBndr]) DCxt DType
           | DDefaultSigD Name DType
           | DPatSynD Name PatSynArgs DPatSynDir DPat
           | DPatSynSigD Name DPatSynType

--- a/Language/Haskell/TH/Desugar/Core.hs
+++ b/Language/Haskell/TH/Desugar/Core.hs
@@ -709,10 +709,10 @@ dsDec (ClassD cxt n tvbs fds decs) =
                      <*> pure fds <*> dsDecs decs)
 #if __GLASGOW_HASKELL__ >= 711
 dsDec (InstanceD over cxt ty decs) =
-  (:[]) <$> (DInstanceD <$> pure over <*> dsCxt cxt <*> dsType ty <*> dsDecs decs)
+  (:[]) <$> (DInstanceD over Nothing <$> dsCxt cxt <*> dsType ty <*> dsDecs decs)
 #else
 dsDec (InstanceD cxt ty decs) =
-  (:[]) <$> (DInstanceD <$> pure Nothing <*> dsCxt cxt <*> dsType ty <*> dsDecs decs)
+  (:[]) <$> (DInstanceD Nothing Nothing <$> dsCxt cxt <*> dsType ty <*> dsDecs decs)
 #endif
 dsDec d@(SigD {}) = dsTopLevelLetDec d
 dsDec (ForeignD f) = (:[]) <$> (DForeignD <$> dsForeign f)
@@ -781,10 +781,11 @@ dsDec (PatSynD n args dir pat) = do
   return [DPatSynD n args dir' pat']
 dsDec (PatSynSigD n ty) = (:[]) <$> (DPatSynSigD n <$> dsType ty)
 dsDec (StandaloneDerivD mds cxt ty) =
-  (:[]) <$> (DStandaloneDerivD <$> mapM dsDerivStrategy mds <*> dsCxt cxt <*> dsType ty)
+  (:[]) <$> (DStandaloneDerivD <$> mapM dsDerivStrategy mds
+                               <*> pure Nothing <*> dsCxt cxt <*> dsType ty)
 #else
 dsDec (StandaloneDerivD cxt ty) =
-  (:[]) <$> (DStandaloneDerivD Nothing <$> dsCxt cxt <*> dsType ty)
+  (:[]) <$> (DStandaloneDerivD Nothing Nothing <$> dsCxt cxt <*> dsType ty)
 #endif
 dsDec (DefaultSigD n ty) = (:[]) <$> (DDefaultSigD n <$> dsType ty)
 #endif

--- a/Language/Haskell/TH/Desugar/Sweeten.hs
+++ b/Language/Haskell/TH/Desugar/Sweeten.hs
@@ -116,6 +116,7 @@ decToTH (DInstanceD over mtvbs _cxt _ty decs) =
 #if __GLASGOW_HASKELL__ < 800 || __GLASGOW_HASKELL__ >= 802
                                 ([], DForallT _tvbs _cxt _ty)
 #else
+                                -- See #117
                                 error $ "Explicit foralls in instance declarations "
                                      ++ "are broken on GHC 8.0."
 #endif
@@ -173,6 +174,7 @@ decToTH (DStandaloneDerivD mds mtvbs _cxt _ty) =
 #if __GLASGOW_HASKELL__ < 710 || __GLASGOW_HASKELL__ >= 802
                                 ([], DForallT _tvbs _cxt _ty)
 #else
+                                -- See #117
                                 error $ "Explicit foralls in standalone deriving declarations "
                                      ++ "are broken on GHC 7.10 and 8.0."
 #endif

--- a/Language/Haskell/TH/Desugar/Sweeten.hs
+++ b/Language/Haskell/TH/Desugar/Sweeten.hs
@@ -107,12 +107,17 @@ decToTH (DDataD Newtype cxt n tvbs _mk [con] derivings) =
 decToTH (DTySynD n tvbs ty) = [TySynD n (map tvbToTH tvbs) (typeToTH ty)]
 decToTH (DClassD cxt n tvbs fds decs) =
   [ClassD (cxtToTH cxt) n (map tvbToTH tvbs) fds (decsToTH decs)]
-#if __GLASGOW_HASKELL__ >= 711
-decToTH (DInstanceD over cxt ty decs) =
-  [InstanceD over (cxtToTH cxt) (typeToTH ty) (decsToTH decs)]
+decToTH (DInstanceD over mtvbs _cxt _ty decs) =
+  [instanceDToTH over cxt' ty' decs]
+  where
+  (cxt', ty') = case mtvbs of
+                  Nothing    -> (_cxt, _ty)
+                  Just _tvbs ->
+#if __GLASGOW_HASKELL__ < 800 || __GLASGOW_HASKELL__ >= 802
+                                ([], DForallT _tvbs _cxt _ty)
 #else
-decToTH (DInstanceD _ cxt ty decs) =
-  [InstanceD (cxtToTH cxt) (typeToTH ty) (decsToTH decs)]
+                                error $ "Explicit foralls in instance declarations "
+                                     ++ "are broken on GHC 8.0."
 #endif
 decToTH (DForeignD f) = [ForeignD (foreignToTH f)]
 #if __GLASGOW_HASKELL__ > 710
@@ -159,18 +164,22 @@ decToTH (DClosedTypeFamilyD (DTypeFamilyHead n tvbs frs _ann) eqns) =
 #endif
 decToTH (DRoleAnnotD n roles) = [RoleAnnotD n roles]
 #endif
+decToTH (DStandaloneDerivD mds mtvbs _cxt _ty) =
+  [standaloneDerivDToTH mds cxt' ty']
+  where
+  (cxt', ty') = case mtvbs of
+                  Nothing    -> (_cxt, _ty)
+                  Just _tvbs ->
+#if __GLASGOW_HASKELL__ < 710 || __GLASGOW_HASKELL__ >= 802
+                                ([], DForallT _tvbs _cxt _ty)
+#else
+                                error $ "Explicit foralls in standalone deriving declarations "
+                                     ++ "are broken on GHC 7.10 and 8.0."
+#endif
 #if __GLASGOW_HASKELL__ < 709
-decToTH (DStandaloneDerivD {}) =
-  error "Standalone deriving supported only in GHC 7.10+"
 decToTH (DDefaultSigD {})      =
   error "Default method signatures supported only in GHC 7.10+"
 #else
-decToTH (DStandaloneDerivD _mds cxt ty) =
-  [StandaloneDerivD
-#if __GLASGOW_HASKELL__ >= 801
-    (fmap derivStrategyToTH _mds)
-#endif
-    (cxtToTH cxt) (typeToTH ty)]
 decToTH (DDefaultSigD n ty)        = [DefaultSigD n (typeToTH ty)]
 #endif
 #if __GLASGOW_HASKELL__ >= 801
@@ -323,6 +332,26 @@ conToTH (DCon tvbs cxt n fields rty)
 
     con' :: Con
     con' = conToTH $ DCon [] [] n fields rty
+#endif
+
+instanceDToTH :: Maybe Overlap -> DCxt -> DType -> [DDec] -> Dec
+instanceDToTH _over cxt ty decs =
+  InstanceD
+#if __GLASGOW_HASKELL__ >= 800
+            _over
+#endif
+            (cxtToTH cxt) (typeToTH ty) (decsToTH decs)
+
+standaloneDerivDToTH :: Maybe DDerivStrategy -> DCxt -> DType -> Dec
+#if __GLASGOW_HASKELL__ >= 710
+standaloneDerivDToTH _mds cxt ty =
+  StandaloneDerivD
+#if __GLASGOW_HASKELL__ >= 802
+                   (fmap derivStrategyToTH _mds)
+#endif
+                   (cxtToTH cxt) (typeToTH ty)
+#else
+standaloneDerivDToTH _ _ _ = error "Standalone deriving supported only in GHC 7.10+"
 #endif
 
 foreignToTH :: DForeign -> Foreign

--- a/Test/Dec.hs
+++ b/Test/Dec.hs
@@ -12,6 +12,9 @@ rae@cs.brynmawr.edu
 #if __GLASGOW_HASKELL__ >= 707
 {-# LANGUAGE RoleAnnotations #-}
 #endif
+#if __GLASGOW_HASKELL__ >= 710
+{-# LANGUAGE DeriveAnyClass #-}
+#endif
 
 {-# OPTIONS_GHC -fno-warn-orphans -fno-warn-name-shadowing #-}
 
@@ -43,6 +46,13 @@ $(S.dectest14)
 
 #if __GLASGOW_HASKELL__ >= 710
 $(S.dectest15)
+#endif
+
+#if __GLASGOW_HASKELL__ < 800 || __GLASGOW_HASKELL__ >= 802
+$(S.dectest16)
+#endif
+#if __GLASGOW_HASKELL__ >= 802
+$(S.dectest17)
 #endif
 
 $(fmap unqualify S.instance_test)

--- a/Test/DsDec.hs
+++ b/Test/DsDec.hs
@@ -12,6 +12,9 @@ rae@cs.brynmawr.edu
 #if __GLASGOW_HASKELL__ >= 707
 {-# LANGUAGE RoleAnnotations #-}
 #endif
+#if __GLASGOW_HASKELL__ >= 710
+{-# LANGUAGE DeriveAnyClass #-}
+#endif
 #if __GLASGOW_HASKELL__ >= 801
 {-# LANGUAGE DerivingStrategies #-}
 #endif
@@ -72,6 +75,13 @@ $(dsDecSplice S.dectest14)
 
 #if __GLASGOW_HASKELL__ >= 710
 $(dsDecSplice S.dectest15)
+#endif
+
+#if __GLASGOW_HASKELL__ < 800 || __GLASGOW_HASKELL__ >= 802
+$(return $ decsToTH [S.ds_dectest16])
+#endif
+#if __GLASGOW_HASKELL__ >= 802
+$(return $ decsToTH [S.ds_dectest17])
 #endif
 
 $(do decs <- S.rec_sel_test

--- a/Test/Splices.hs
+++ b/Test/Splices.hs
@@ -423,6 +423,35 @@ dectest15 = [d| infixl 5 :**:, :&&:, :^^:, `ActuallyPrefix`
                   (:^^:) :: Int -> Int -> Int -> InfixGADT Int
                   (:!!:) :: Char -> Char -> InfixGADT Char |]
 
+class ExCls a
+data ExData1 a
+data ExData2 a
+
+ds_dectest16 = DInstanceD Nothing (Just [DPlainTV (mkName "a")]) []
+                (DConT ''ExCls `DAppT`
+                  (DConT ''ExData1 `DAppT` DVarT (mkName "a"))) []
+dectest16 :: Q [Dec]
+dectest16 = return [ InstanceD
+#if __GLASGOW_HASKELL__ >= 800
+                       Nothing
+#endif
+                       [] (ForallT [PlainTV (mkName "a")] []
+                                   (ConT ''ExCls `AppT`
+                                     (ConT ''ExData1 `AppT` VarT (mkName "a")))) [] ]
+ds_dectest17 = DStandaloneDerivD Nothing (Just [DPlainTV (mkName "a")]) []
+                (DConT ''ExCls `DAppT`
+                  (DConT ''ExData2 `DAppT` DVarT (mkName "a")))
+#if __GLASGOW_HASKELL__ >= 710
+dectest17 :: Q [Dec]
+dectest17 = return [ StandaloneDerivD
+#if __GLASGOW_HASKELL__ >= 802
+                       Nothing
+#endif
+                       [] (ForallT [PlainTV (mkName "a")] []
+                                   (ConT ''ExCls `AppT`
+                                     (ConT ''ExData2 `AppT` VarT (mkName "a")))) ]
+#endif
+
 instance_test = [d| instance (Show a, Show b) => Show (a -> b) where
                        show _ = "function" |]
 

